build/test/TestAssetProjects.targets

```diff
@@ -0,0 +1,62 @@
+<?xml version="1.0" encoding="utf-8"?>
+<Project ToolsVersion="14.0" DefaultTargets="Test" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <Target Name="SetupTestAssetProjectData" 
+          DependsOnTargets="AssociateTestAssetInputFilesWithProject">                          
+  </Target>
+
+  <Target Name="AssociateTestAssetInputFilesWithProject" 
+          DependsOnTargets="IdentifyTestAssetInputFilesForProject;" 
+          Outputs="%(TestAssetProjects.ProjectPath)">
+    <PropertyGroup>
+      <CurrentProject>%(TestAssetProjects.ProjectPath)</CurrentProject>
+    </PropertyGroup>
+
+    <ItemGroup>
+      <TestAssetProjects Condition=" '%(ProjectPath)' == '$(CurrentProject)' ">
+        <BuildInputs>@(TestAssetProjectInputs)</BuildInputs>
+      </TestAssetProjects>
+    </ItemGroup>
+  </Target>
+
+  <Target Name="IdentifyTestAssetInputFilesForProject" 
+          DependsOnTargets="SetupBuildTestAssetProjectInputs;" 
+          Outputs="%(TestAssetProjects.ProjectPath)">
+    <ItemGroup>
+      <TestAssetProjectInputs Include="%(TestAssetProjects.BuildInputIncludeFilter)" 
+                              Exclude="%(TestAssetProjects.BuildInputExcludeFilter)">
+        <ProjectPath>%(TestAssetProjects.ProjectPath)</ProjectPath>
+      </TestAssetProjectInputs>
+    </ItemGroup>
+  </Target>
+
+  <Target Name="SetupBuildTestAssetProjectInputs">
+    <PropertyGroup>
+      <PathSeparator>/</PathSeparator>
+      <PathSeparator Condition=" '$(OSName)' == 'win' ">\</PathSeparator>
+    </PropertyGroup>
+
+    <ItemGroup>
+      <NoAutoBuildTestAssets Include="TestAssets$(PathSeparator)TestProjects$(PathSeparator)**$(PathSeparator).noautobuild" />
+      <NoAutoBuildTestAssetProjects Include="%(NoAutoBuildTestAssets.RelativeDir)project.json" />
+
+      <TestAssetProjects Include="TestAssets$(PathSeparator)TestProjects$(PathSeparator)**$(PathSeparator)project.json" 
+                         Exclude="@(NoAutoBuildTestAssetProjects)"/>
+      <TestAssetProjects>
+        <BuildInputIncludeFilter>%(RelativeDir)**/*.*</BuildInputIncludeFilter>
+        <BuildInputExcludeFilter>%(RelativeDir)bin/**/*.*;%(RelativeDir)obj/**/*.*</BuildInputExcludeFilter>
+        <OutputDir>$([System.IO.Directory]::GetParent(%(Identity)))</OutputDir>
+        <ProjectPath>%(Identity)</ProjectPath>
+      </TestAssetProjects>
+
+      <TestAssetProjects>
+        <OutputName>$([System.IO.Path]::GetFileName(%(OutputDir)))</OutputName>
+        <OutputName Condition=" '%(RelativeDir)' == 'TestAssets/TestProjects/AppWithOutputAssemblyName/' ">MyApp</OutputName>
```

Comment:
I could create a seperate itemgroup, OutputNameOverrides, and then merge them. I chose this approach because there are less MSBuild concepts in play. What are your thoughts on the alternative?